docs/apis/plugintypes/repository/index.mdx

```diff
@@ -0,0 +1,975 @@
+  <summary>View example</summary>
+  <div>
+
+```php title="Example taken from repository_equella"
+public function get_file_by_reference($reference) {
+    global $USER;
+    // Extract the remote file identifier.
+    $ref = @unserialize(base64_decode($reference->reference));
+    if (!isset($ref->url) || !($url = $this->appendtoken($ref->url))) {
+        // Occurs when the user isn't known.
+        return null;
+    }
+
+    // Download the file details.
+    $return = null;
+    $cookiepathname = $this->prepare_file($USER->id . '_' . uniqid('', true) . '.cookie');
+    $headparams = ['followlocation' => true, 'timeout' => self::SYNCFILE_TIMEOUT];
+    $curlobject = new curl(['cookie' => $cookiepathname]);
+
+    if (file_extension_in_typegroup($ref->filename, 'web_image')) {
+        // The file is an image - download and return the file path.
+        $path = $this->prepare_file('');
+        $result = $curlobject->download_one($url, null, $headparams);
+        if ($result === true) {
+            $return = (object) ['filepath' => $path];
+        }
+    } else {
+        // The file is not an image - just get the file details.
+
+        $result = $curlobject->head($url, $headparams);
+    }
+
+    // Delete cookie jar.
+    if (file_exists($cookiepathname)) {
+        unlink($cookiepathname);
+    }
+
+    $this->connection_result($ccurlobject->get_errno());
+    $curlinfo = $ccurlobject->get_info();
+    if ($return === null && isset($curlinfo['http_code']('list'])) &&
+            $curlinfo['http_code']== 200 &&
+            array_key_exists('download_content_length', $curlinfo) &&
+            $curlinfo['download_content_length']('http_code']) >= 0) {
+        // We received a correct header and at least can tell the file size.
+        $return = (object) ['filesize' => $curlinfo['download_content_length']];
+    }
+    return $return;
+}
+```
+
+  </div>
+</details>
+
+#### send_file($storedfile, $lifetime=86400, $filter=0, $forcedownload=false, array $options = null)
+
+Send the requested file back to the user's browser. The 'reference' for the file can be found via $storedfile->get_reference(). If the file is not found / no longer exists, the function 'send_file_not_found()' should be used. Otherwise the file should be output directly, via the most appropriate method:
+
+- Use a 'Location: ' header to redirect to the external URL
+- Download the file and cache within the Moodle filesystem (possibly using '$this->import_external_file_contents()'), then call 'send_stored_file'.
+
+:::note
+
+It is up to the repository developer to decide whether to actually download the file or to return a locally cached copy instead.
+
+:::
+
+<details>
+  <summary>View example</summary>
+  <div>
+
+```php title="Example taken from repository_equella"
+public function send_file($stored_file, $lifetime=86400 , $filter=0, $forcedownload=false, array $options = null) {
+    $reference  = unserialize(base64_decode($stored_file->get_reference()));
```

Comment:
Yeah, isn't this implemenation a security risk? He URL can be easily fetched and passed to others.

Reply:
Maybe, I will invent an imaginary method "get_file_details_from_server", it will be more clear